tests/import_specifications/test_individual_parsers.py

```diff
@@ -0,0 +1,235 @@
+):
+    s = sep
+    input_ = temp_dir / str(uuid.uuid4())
+    with open(input_, "w") as test_file:
+        test_file.writelines([
+            "Data type: other_type; Version: 1\n",
+            f"spec1{s} spec2{s} spec3{s} spec4\n",
+            f"Spec 1{s} Spec 2{s} Spec 3{s} Spec 4\n",
+            f"val1 {s} val2{s}    7     {s} 3.2\n",
+            f"val3 {s} val4{s} 1{s} 8.9\n",
+            f"val5 {s} val6{s} int{s} float\n",
+        ])
+
+    res = parser(input_)
+
+    assert res == ParseResults(frozendict(
+        {"other_type": ParseResult(SpecificationSource(input_),
+            tuple([
+                frozendict({"spec1": "val1", "spec2": "val2", "spec3": "7", "spec4": "3.2"}),
+                frozendict({"spec1": "val3", "spec2": "val4", "spec3": "1", "spec4": "8.9"}),
+                frozendict({"spec1": "val5", "spec2": "val6", "spec3": "int", "spec4": "float"}),
+            ])
+        )}
+    ))
+
+def test_xsv_parse_fail_no_file(temp_dir: Path):
+    input_ = temp_dir / str(uuid.uuid4())
+    with open(input_, "w") as test_file:
+        test_file.writelines([
+            "Data type: other_type; Version: 1\n",
+            "spec1, spec2, spec3, spec4\n",
+            "Spec 1, Spec 2, Spec 3, Spec 4\n",
+            "val1 , val2,    7     , 3.2\n",
+        ])
+    input_ = Path(str(input_)[-1])
+
+    res = parse_csv(input_)
+
+    assert res == ParseResults(errors=tuple([
+        Error(ErrorType.FILE_NOT_FOUND, source_1=SpecificationSource(input_))
+    ]))
+
+
+def _xsv_parse_fail(
+    temp_dir: Path, lines: list[str], parser: Callable[[Path], ParseResults], message: str
+):
+    input_ = temp_dir / str(uuid.uuid4())
+    with open(input_, "w") as test_file:
+        test_file.writelines(lines)
+
+    res = parser(input_)
+    expected = ParseResults(errors=tuple([Error(
+        ErrorType.PARSE_FAIL, message, SpecificationSource(input_)
+    )]))
+    assert res == expected
+
+
+def test_xsv_parse_fail_empty_file(temp_dir: Path):
+    _xsv_parse_fail(temp_dir, [], parse_csv, "Missing data type / version header")
+
+
+def test_xsv_parse_fail_bad_datatype_header(temp_dir: Path):
+    err = ('Invalid header; got "This is the wrong header", expected "Data type: '
+        + '<data_type>; Version: <version>"')
+    _xsv_parse_fail(temp_dir, ["This is the wrong header"], parse_csv, err)
+
+
+def test_xsv_parse_fail_bad_version(temp_dir: Path):
+    err = "Schema version 87 is larger than maximum processable version 1"
+    _xsv_parse_fail(temp_dir, ["Data type: foo; Version: 87"], parse_csv, err)
+
+def test_xsv_parse_fail_missing_column_headers(temp_dir: Path):
+    err = "Expected 2 column header rows"
+    _xsv_parse_fail(temp_dir, ["Data type: foo; Version: 1\n"], parse_csv, err)
+
+    lines = ["Data type: foo; Version: 1\n", "head1, head2\n"]
+    _xsv_parse_fail(temp_dir, lines, parse_csv, err)
+
+
+def test_xsv_parse_fail_missing_data(temp_dir: Path):
+    err = "No data in file"
```

Comment:
I wonder if it would be better to say something about headers vs content, because there is data in the file, it just happens that its header data rather than body data.

Reply:
How would you phrase it?

Reply:
It's probably a rare error, but it this error might occur when a user downloads the template and then uploads a blank template instead of filling it out.

Some ideas could be

Import Spec Error:
Template body missing
Template body blank
Only import headers provided, import body missing
No data in file: (Headers found, body missing)
No data in file: Please fill out the import specification template for <app>

Reply:
Updated